File: c_version/simulate.cpp
#include <math.h>
#include <iostream>

#include "constants.hpp"
#include "satellite.hpp"
#include "utils.hpp"

#define RISE_EVENT 0
#define SET_EVENT 1

void analyse_timeline(const int nb_sat_ini, const std::vector<int> event_types, const std::vector<double> event_dates, const double total_sim_time, double *t_blind, int *nsat_max)
{
    int nsat, sorted_index;
    int ns;
    double last_blind_date;
    std::vector<size_t> indices;
    indices = argsort(event_dates);

    ns = event_dates.size();
    nsat = nb_sat_ini;
    *t_blind = 0;
    *nsat_max = 0;
    last_blind_date = 0;
    for (int i = 0; i < ns; i++)
    {
        sorted_index = indices[i];

        if (event_types[sorted_index] == RISE_EVENT)
        {
            if (nsat == 0)
                *t_blind += event_dates[sorted_index] - last_blind_date;
            nsat += 1;
            std::cout << event_dates[sorted_index] << " rise, nsat=" << nsat << ", t_blind=" << *t_blind << std::endl;
        }
        else if (event_types[sorted_index] == SET_EVENT)
        {
            nsat -= 1;
            if (nsat == 0)
                last_blind_date = event_dates[sorted_index];
            else if (nsat < 0) {
                std::cerr << "nsat:" << nsat << std::endl;
                throw;
            }
            std::cout << event_dates[sorted_index] << " set, nsat=" << nsat << ", t_blind=" << *t_blind << std::endl;
        }
        else {
            std::cerr << "unknown event:" << event_types[sorted_index]  << std::endl;
            throw;
        }

        if (*nsat_max < nsat)
            *nsat_max = nsat;
    }

    if (*nsat_max == 0)
        *t_blind = total_sim_time;
}

void simulate(double lat, double inc, int nsat, int npla, int pha, double alt_km, double *t_blind, int *nsat_max)
{
    std::vector<int> event_types;
    std::vector<double> event_dates;
    int s = nsat / npla;
    int status;
    int satellite_id, nb_sat;
    double elev_mask;
    Satellite sat;
    event_type events;
    VectorXd obs(6), pv_sat(6);
    obs = llavpa_to_itrf(0, lat, 0, 0, 0, 0);

    const double sma = Req + alt_km * 1e3;
    const double px_dbm = 36;
    const double nf_db = 3;
    const double eta = 0.7;
    const double fcarrier_mhz = 2000;
    const double cn0_lim_dbhz = 46.46595211;
    const double alpha = 0.0305;
    const double t0_k = 290;
    // const double t_sim = 5 * 86400;
    const double t_sim = 20000;
    const double firstraan = 0.;
    const double argp = 0.0;
    double raan, meanAnomaly, t_start;

    nb_sat = 0;
    for (int idxP = 0; idxP < npla; idxP++)
    {
        raan = firstraan * 180 / M_PI + idxP * 360.0 / npla;
        for (int idxS = 0; idxS < s; idxS++)
        {
            meanAnomaly = idxP * pha * 360.0 / nsat + idxS * 360.0 / s;
            elev_mask = compute_elevation_mask(px_dbm, nf_db, eta, fcarrier_mhz, cn0_lim_dbhz, alpha, t0_k, alt_km);

            sat = Satellite(
                sma,
                inc,
                argp,
                meanAnomaly * M_PI / 180,
                raan * M_PI / 180);

            t_start = 0;
            while (true)
            {
                sat.find_events(obs, t_start, elev_mask, &events);
                t_start = events.t_culmination + events.tup_max;
                std::cout << events.t_culmination << ", " << events.t_rise << ", " << events.is_initially_visible << std::endl;
                if (events.is_initially_visible)
                    nb_sat++;
                if (events.t_culmination > t_sim)
                    break;
                if (events.t_rise > 0)
                {
                    event_types.push_back(RISE_EVENT);
                    event_dates.push_back(events.t_rise);
                    event_types.push_back(SET_EVENT);
                    event_dates.push_back(events.t_set);
                }
            }
        }
    }

    std::cout << event_dates.size() << std::endl;
    analyse_timeline(nb_sat, event_types, event_dates, t_sim, t_blind, nsat_max);
}

File: src/constellation_design/simulate.py
from datetime import datetime, timezone
from dataclasses import dataclass
from enum import Enum

import numpy as np
from numpy import arcsin, pi, log10
from numpy.polynomial import Polynomial
from blocksim.satellite.Satellite import (
    CircleSatellite,
    generateWalkerDeltaConstellation,
)
from blocksim.utils import llavpa_to_itrf, itrf_to_azeld
from blocksim.constants import Req, kb, c as clum


class EventType(Enum):
    RISE = "rise"
    SET = "set"


@dataclass
class Event:
    satellite: int
    type: EventType
    date: float


def analyse_timeline(init, events, total_sim_time):
    tl = []
    for ksat, sat in enumerate(events):
        for e in sat:
            tl.append(Event(satellite=ksat, type=EventType.RISE, date=e["rise"]))
            tl.append(Event(satellite=ksat, type=EventType.SET, date=e["set"]))
    tl.sort(key=lambda x: x.date)

    nsat = len(np.where(init > 0)[0])
    blind_e = Event(satellite=-1, type=EventType.RISE, date=0)
    t_blind = 0
    e: Event
    nsat_max = nsat
    for e in tl:
        if e.type == EventType.RISE:
            if nsat == 0:
                t_blind += e.date - blind_e.date
            nsat += 1
        else:
            nsat -= 1
            if nsat == 0:
                blind_e = e
            elif nsat < 0:
                raise AssertionError
        if nsat_max < nsat:
            nsat_max = nsat
        print(e,nsat,t_blind)

    if nsat_max == 0:
        t_blind = total_sim_time

    return t_blind, nsat_max


def compute_elevation_mask(alt_km):
    # Parametres du probleme
    # ======================
    px = 36  # dBm
    NF = 3  # dB
    z1_J1 = 1.616339347
    eta = 0.7
    wl = clum / (2e9)
    cn0_lim = 46.46595211
    alpha = 0.0305

    # Calculs des constantes K et Q
    # =============================
    sma = Req + alt_km * 1e3
    K = (
        -36
        + px
        - NF
        + 10
        * log10((eta**2 * wl**2 * z1_J1**4 * sma**4) / (16 * pi**2 * kb * 290 * Req**4))
    )
    Q = 10 ** ((K - cn0_lim) / 20)

    # Calcul de l'élévation
    # =====================
    p = Polynomial(
        [
            alpha,
            2,
            -(
                (2 * alpha**2 - 1) * sma**2
                + (1 - 2 * alpha**2) * Req**2
                + 2 * alpha * Q * Req
                + Q**2
            )
            / (alpha * sma**2 - alpha * Req**2),
            -(4 * alpha * sma**2 - 4 * alpha * Req**2 + 2 * Q * Req)
            / (alpha * sma**2 - alpha * Req**2),
            ((alpha**2 - 2) * sma**2 + (2 - alpha**2) * Req**2 + 2 * alpha * Q * Req)
            / (alpha * sma**2 - alpha * Req**2),
            (2 * alpha * sma**2 - 2 * alpha * Req**2 + 2 * Q * Req)
            / (alpha * sma**2 - alpha * Req**2),
            1 / alpha,
        ]
    )

    rts = p.roots()
    r0 = rts[np.where((rts > arcsin(alpha)) & (rts < 1) & (np.abs(np.imag(rts)) < 1e-9))[0]]
    assert len(r0) == 1
    elev_mask = arcsin(r0[0])

    return elev_mask


def simulate(lat, inc, nsat, npla, pha, alt_km):
    elev_mask = compute_elevation_mask(alt_km)

    t0 = datetime(2023, 6, 27, 12, 0, 0, tzinfo=timezone.utc)
    firstraan = 0.0
    lon = 0.0
    # tps_max = 5 * 86400
    tps_max = 20000
    sma = Req + alt_km * 1e3

    satellites = generateWalkerDeltaConstellation(
        "sim", sma, inc, firstraan, nsat, npla, pha, t0, prop=CircleSatellite
    )
    obs = llavpa_to_itrf((lon, lat, 0, 0, 0, 0))
    events = list()
    init = list()
    sat: CircleSatellite
    for sat in satellites:
        pv_sat = sat.getGeocentricITRFPositionAt(0)
        _, el0, _, _, _, _ = itrf_to_azeld(obs, pv_sat)
        init.append(el0 - elev_mask)
        events.append(
            sat.find_events(obs, t0=0, t1=tps_max, elevation=elev_mask),
        )

    t_blind, nsat_max = analyse_timeline(np.array(init), events, tps_max)

    return t_blind, nsat_max, tps_max, elev_mask
